### PR TITLE
Added new, more flexible script for parsing network interfaces

### PR DESCRIPTION
**Pull request by roobre**
````
Usage: ./scripts/iface-ng [-d down] [-n "no ip"] [-c FF0000] [-C 00FF00] [-e] [-p] [-s] [-6]
    -d down: Placeholder text if the specified interface is down. If set to an empty string,
       the entire block is hidden (useful to save space on small screens).
    -n "no ip": Placeholder text to display when the interface is up but it does not
       have any ip address.
    -c FF0000: Text color when interface is down.
    -C 00FF00: Text color when interface is up.
    -e: Show interface even if it does not exist (useful for tun/tap).
    -p: Include CIDR prefix.
    -s: Include link speed (ethernet only).
    -6: Use ipv6 address insead of v4.
````

### CODE REVIEW

**Comment by vivien** [comment]
Hi @roobre!

The network block(s) is a tricky subject! @Nycroth can I ask your feedback on this? I'll give it a try anyway in the meantime.

**Comment by Nycroth** [comment]
Yeah, it's a tough call. The problem @roobre, is that the networking scripts have flip-flopped a bunch since they were first written (#14, #15, #16, #17, #61, #83). Everyone loves something a little different with these scripts.

First it was a single IP address script. Then it was updated to support multiple interfaces and IPv6. Then a wireless script was made. Then they were merged. Then it had things like ESSID, CIDR prefix, signal strength, and bandwidth all added. Then there were too many flags and output got messy (people get particular on even the order elements appear and a format string input would just be ridiculous) so the script was stripped down and split into 3 scripts.

This last change was primarily to remove the flags. The problem with option flags is that they need documenting and they make the script feel like a standalone application instead of a starter template. The only real documentation we provide on scripts is the provided config example, so too many flags becomes a bit tedious to provide examples for. In #61 we informally agreed that scripts used should avoid flags and rely instead on the BLOCK_INSTANCE environment variable to pass configuration. This forces scripts to remain simple and focus on a task instead of features. Some exceptions persist but we felt that scripts should aim to provide a useful and widely used service and avoid attempting to provide everything for everyone. Ideally, they are also simple and easy to follow with clear documentation so that even beginners feel comfortable modifying them to meet their desires.

I like your work and note that many of the features you've provided [colour selection, no-IP text, tunnel persistence] are things others have already asked for. Yet, I would rather your work be written up in a couple wiki entries as patch guides for others who want such functionality. I would even suggest splitting the examples wiki into a section for script examples and a section for core script mods/patches.

Though, as always, I'm open minded.

**Comment by roobre** [comment]
Thanks for the thorough response, @Nycroth. I gotta say you are right. All the flag stuff is messy, but, if you come to think about it, most linux desktop users use linux because of the customization options it provides, and the same thing goes to i3blocks vs i3status, which I stopped using because it wasn't customizable enough.

So yes, the flags are ugly, and yes, we should avoid them. But we need them at the same time to allow more customization. How may we solve this conflict? Well, I just had an idea. 
**What if i3blocks would provide an standard system to pass config/format options to scripts?** I'm just dropping it as a suggestion here, but hey, it could be great. Something like:

````
[iface-ng]
label=wlan0:
instance=wlan0
color=#00FF00
format=%i/%c (%s Mbps)
arbitraryOption=3
interval=5
````

Coming to think about it, it's something i3status already does (most of its segments accept a `format` argument, and seems natural to me.

Passing options to scripts is important imho (customization is the linux desktop motto), so maybe thinking about a standard way of doing it is a nice way to improve i3blocks.

What's your opinion on this subject?

Regards, Roberto.

**Comment by Nycroth** [comment]
Core functionality is more @vivien's specialty. He'll have to weigh in on a format parameter. How would you suggest scripts handle these? Would it be passed as a string to the script in another environment variable or would i3blocks be responsible for some modifications and interfacing?

If the prior, how would you suggest a bash script best handle this (a hypothetical code sample would be appreciated)? Also, if this is the case, nothing would stop it being passed as a part of the instance variable (short of the character limit). As an example, pass `wlan0+l%i/%c (%s Mbps)` in the instance and parse `${BLOCK_INSTANCE#*+}` for the format string (any separator could be used, I used + because the date command does). If the latter, what would the alternative entail?

The other key to a good feature is one which will see much use. Beyond the networking script and maybe media player scripts, are there any others that might utilize such a feature? Date surely, but the most used method is to just call the date command directly.

---

Arbitrary options (as for as I know) would require quite a bit of work with the core. Also, this would mean any typo would result in legal but unexpected behaviour. As an example, being Canadian colour is spelt with a letter U. My editor's dictionary rightly suggests this and I have to correct myself when I code (I spell it with a U out of habit) as the standard in coding is to use the American spelling of all words. This would be obvious as the colour would be the default but typos are quite common. Less enjoyable troubleshooting would be words like interval which leads to the block not being displayed or instance which would lead to a script not behaving as desired by the user. This, though minor, should still be a consideration. And finally, in future all added fields (eg. markup) would inevitably lead to scripts needing to be updated as a once arbitrary parameter is now a i3blocks parameter. An alternative here would be to pass every configuration option to the script regardless of use by i3blocks, so it could be worked around.

---

I agree Linux users are a savvy bunch. However, there are limits each will go through to configure their system. Some enjoy to just KDE their way and use a system to accomplish their goals. Others like i3blocks users prefer to configure everything to the last detail. Obviously our users are in the latter :wink:. 

Here is where every project just has to clearly identify what the project goal is. For example the GO language: "[The goals of the Go project were to eliminate the slowness and clumsiness of software development at Google, and thereby to make the process more productive and scalable. The language was designed by and for people who write—and read and debug and maintain—large software systems](http://talks.golang.org/2012/splash.article)" I only use this because it was in a fun [article I read recently](https://texlution.com/post/why-go-is-doomed-to-succeed/) but it illustrates the point. I'm not exactly sure what i3blocks main goal is but knowing what it is would help sort out decisions going forward.

You're script works extremely well. In fact, it could even become a format string driven networking information application. Networking is a computing geek rabbit hole. The more you learn, the more you're always curious about your network (it's one of my core competencies so I sympathize, lol). I'm just curious about what everyone on this project feels the relationship between the core, the scripts, and the wiki is.

Sorry I'm brambly but I find longer posts better get the reader inside the author's head. Can't wait to read replies.

**Comment by roobre** [comment]
Hi, and thanks for the thorough response again. I'll answer your comments by parts, if you dont mind.

---

> How would you suggest scripts handle these? Would it be passed as a string to the script in another environment variable or would i3blocks be responsible for some modifications and interfacing?

The first option I thought was allowing block to have arbitrary options, say `foo=bar`, that would be passed as `$BLOCK_FOO` verbatim, to allow any text to be passed without i3blocks caring about it.

---

> If the prior, how would you suggest a bash script best handle this (a hypothetical code sample would be appreciated)?

By the same way the `$BLOCK_INSTANCE` variable is handled, this is, reading the `$BLOCK_FOO` variable from the environment.

---

> nothing would stop it being passed as a part of the instance variable (short of the character limit). As an example, pass wlan0+l%i/%c (%s Mbps) in the instance and parse ${BLOCK_INSTANCE#*+}

That's ineed a good idea, as it would work without touching i3blocks core code at all. It has, however, the lack of readability as a con. If you look at the i3blocks config file and you see something like:

````
[mediaplayer]
instance=spotify
format=%t - %a
interval=5
signal=10
````

It's pretty straightforward to figure out what the line `format=%t - %a` does. However, with something like:

````
[mediaplayer]
instance=spotify+%t - %a
interval=5
signal=10
````

It would be a little more obscure.

However, as I've said before, this is a cheaper modification and I wouldn't mind adapting my scripts this way if you decide to make it the standard way of passing format parameters.

---

> Beyond the networking script and maybe media player scripts, are there any others that might utilize such a feature? Date surely, but the most used method is to just call the date command directly.

As far as I've thought, I'd like to be able to specify a format on:

* Network interfaces
* Disk (space being used, space left, % being used, % free)
* Memory (same as disk)
* System load (1min average, 3m, 5m, etc...)
* Battery (Current percentage, current status (CHR/DIS), time left)

But this may be exaggerated, 

---

> Networking is a computing geek rabbit hole. The more you learn, the more you're always curious about your network (it's one of my core competencies so I sympathize, lol).

You're absolutely right, networing is a really fun and absorbing part of this world of ours :3

---

> My feeling from working with this project is that the core is designed to be a lean mean interface from STDOUT to i3blocks.JSON. A sort of "less is more" UNIX approach. The included scripts were then there to provide the most often desired functionality to prevent everyone needing to implement their own. At the same time these scripts would serve as a starting point and example of what you as the user can do when writing your own (prevent the blank canvas problem). Then contrib was merged and I've been a little confused ever since. I've always wondered if the wiki should be for all the one-liners and tips/tricks or if there really are supposed to be links to full scripts in other repos/gists. I personally figured after the merge of contrib into scripts all of the full scripts from the wiki would be merged also but that hasn't happened. Actually, I felt this about the contrib before the merge; I digress.

I agree with the UNIX philosophy, providing an interface between the user-made scripts and the i3bar json. That's why I think it should be more permissive with the way of passing arguments to scripts.

Another approach, btw, would be using long options instead of environment variables. Something like:

````
[mediaplayer]
instance=spotify
format=%t - %a
interval=5
signal=10
````

being translated into

    /usr/libexec/i3blocks/mediaplayer --format="%t - %a"

However, the main point is if we should allow arbitrary options to be passed to the scripts or not, how to implement it is a matter of further discussion, in my opinion.

---

To sum up everything being said, I think we should agree into an standard way of passing arbitrary parameters (or at least one for the format) to scripts. Either via the `$BLOCK_INSTANCE` variable with the `+` style (`instance=spotify+%t - %a`), or using arbitrary directives in the block configurations, which would be passed to the script either as a long option or as an environment variable.

Of course I'll respect whichever you decide, and adapt my scripts accordingly :D

Best regards, Roberto.

**Comment by Nycroth** [comment]
Love the sectioned repose. Thanks!

Hmmm... There's some merit to that last idea of converting from config options to command options. :+1: I would suggest a standard like `param_foo=bar` --> `--foo=bar`. It would keep configs very clean, keep the core code simple, but allow for many more options. I think single arguments `-a -b -c` should still be added directly but making a param converter wouldn't be much work at all and would aid legibility (and support existing programs).

On the matter of the networking script, I wouldn't be opposed to a fourth script that provided a more through availability of networking information. It would have to be format string controlled to be simple and maximally flexible, however. Obviously add it to the existing networking scripts instead of replacing them. The core 3 provide for 90% but another could help that remaining 10%. This way, going forward new networking information readouts would be accepted into the fourth script without impacting the slim nature of the other 3.

Other's thoughts?

**Comment by Nycroth** [comment]
You may want to see and weigh in on #118 about a possible change to how scripts are handled. I'd value your input!
